daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -902,6 +903,105 @@ object SBuiltin {
+                case Some(k) =>
+                  fields(values.size) = k
+                  val _ = values.add(go(v))
+              }
+          }
+          SRecord(id, fields, values)
+        case V.ValueRecord(None, _) =>
+          crash("SValue.fromValue: record missing identifier")
+        case V.ValueStruct(fs) =>
+          val fields = Name.Array.ofDim(fs.length)
+          val values = new util.ArrayList[SValue](fields.length)
+          fs.foreach {
+            case (k, v) =>
+              fields(values.size) = k
+              val _ = values.add(go(v))
+          }
+          SStruct(fields, values)
+        case V.ValueVariant(None, _variant @ _, _value @ _) =>
+          crash("SValue.fromValue: variant without identifier")
+        case V.ValueEnum(None, constructor @ _) =>
+          crash("SValue.fromValue: enum without identifier")
+        case V.ValueOptional(mbV) =>
+          SOptional(mbV.map(go))
+        case V.ValueTextMap(map) =>
+          STextMap(map.mapValue(go).toHashMap)
+        case V.ValueGenMap(entries) =>
+          SGenMap(
+            entries.iterator.map { case (k, v) => go(k) -> go(v) }
+          )
+        case V.ValueVariant(Some(id), variant, value) =>
+          machine.compiledPackages.getPackage(id.packageId) match {
+            case Some(pkg) =>
+              pkg.lookupIdentifier(id.qualifiedName).fold(crash, identity) match {
+                case Ast.DDataType(_, _, data: Ast.DataVariant) =>
+                  SVariant(id, variant, data.constructorRank(variant), go(value))
+                case _ =>
+                  crash(s"definition for variant $id not found")
+              }
+            case None =>
+              throw SpeedyHungry(
+                SResultNeedPackage(
+                  id.packageId,
+                  pkg => {
+                    machine.compiledPackages = pkg
+                    machine.ctrl = translateValue(machine, value)
+                  }
+                ))
+          }
+        case V.ValueEnum(Some(id), constructor) =>
+          machine.compiledPackages.getPackage(id.packageId) match {
+            case Some(pkg) =>
+              pkg.lookupIdentifier(id.qualifiedName).fold(crash, identity) match {
+                case Ast.DDataType(_, _, data: Ast.DataEnum) =>
+                  SEnum(id, constructor, data.constructorRank(constructor))
+                case _ =>
+                  crash(s"definition for variant $id not found")
+              }
+            case None =>
+              throw SpeedyHungry(
+                SResultNeedPackage(
+                  id.packageId,
+                  pkg => {
+                    machine.compiledPackages = pkg
+                    machine.ctrl = translateValue(machine, value)
+                  }
+                ))
+          }
+      }
+    CtrlValue(go(value))
+  }
+
+  final case class SBTranslateValue(value: V[V.ContractId]) extends SBuiltin(0) {
```

Comment:
It looks like this is only used for translating the input value of a DAML script. Why don’t we just call `translateValue` directly for that (similarly to how we called `fromValue` before) instead of adding a new builtin that isn’t used anywhere else?

Reply:
The problem is that `translateValue` may throw exceptions to ask for the package. 
I do not know how to handle it. 
If you some idea, or if you want of fix you are welcome.

Reply:
It only throws an exception if a package is not already in `compiledPackages` right? That will never happen for the script runner so we should be fine to just call it.

Reply:
:+1: